app/components/task-selection/component.js

```diff
@@ -27,215 +33,210 @@ export default Component.extend({
 
-  /**
-   * All available customers
-   *
-   * @property {Customer[]} _customers
-   * @private
-   */
-  @alias('customers')
-  _customers: null,
+      /* istanbul ignore next */
+      customers
+        .then((data) => {
+          let re = new RegExp(`.*${search}.*`, 'i')
 
-  /**
-   * All available projects filtered by the currently selected customer
-   *
-   * @property {Project[]} _projects
-   * @private
-   */
-  @computed('projects.[]', 'customer.id')
-  _projects(projects, id) {
-    if (!id) {
-      return []
+          return asyncResults(data.filter((i) => re.test(i.get('name'))))
+        })
+        .catch(() => {
+          return asyncResults([])
+        })
     }
-
-    return projects.filterBy('customer.id', id)
   },
 
   /**
-   * All available tasks filtered by the currently selected project
+   * Source function for the project autocomplete component
    *
-   * @property {Task[]} _tasks
-   * @private
+   * This either takes the last fetched projects and filters them, or triggers
+   * the call first if none was triggered before.
+   *
+   * @property {Function} projectSource
+   * @public
    */
-  @computed('tasks.[]', 'project.id')
-  _tasks(tasks, id) {
-    if (!id) {
-      return []
-    }
+  @computed
+  projectSource() {
+    return (search, syncResults, asyncResults) => {
+      let fn = this.get('tracking.filterProjects')
+
+      let projects = fn.get('last') || fn.perform(this.get('customer.id'))
 
-    return tasks.filterBy('project.id', id)
+      /* istanbul ignore next */
+      projects
+        .then((data) => {
+          let re = new RegExp(`.*${search}.*`, 'i')
+
+          return asyncResults(data.filter((i) => re.test(i.get('name'))))
+        })
+        .catch(() => {
+          return asyncResults([])
+        })
+    }
   },
 
   /**
-   * The actions of the task selection component
+   * Source function for the task autocomplete component
+   *
+   * This either takes the last fetched projects and filters them, or triggers
+   * the call first if none was triggered before.
    *
-   * @property {Object} actions
+   * @property {Function} taskSource
    * @public
    */
-  actions: {
-    /**
-     * Change action
-     *
-     * This is triggered when the task is selected, not when the customer or
-     * project is selected
-     *
-     * @method change
-     * @param {Task} task The selected task
-     * @public
-     */
-    change(task) {
-      this.get('attrs.on-change')(task)
+  @computed
+  taskSource() {
+    return (search, syncResults, asyncResults) => {
+      let fn = this.get('tracking.filterTasks')
+
+      let tasks = fn.get('last') || fn.perform(this.get('project.id'))
+
+      /* istanbul ignore next */
+      tasks
+        .then((data) => {
+          let re = new RegExp(`.*${search}.*`, 'i')
+
+          return asyncResults(data.filter((i) => re.test(i.get('name'))))
+        })
+        .catch(() => {
+          return asyncResults([])
+        })
```

Comment:
I already considered this one but decided not to extract it because we would need to pass `asyncResults`. "Duplication" is the lesser evil IMHO so I'll let this be.